injector/network_disruption.go

```diff
@@ -374,78 +416,442 @@ func (i *networkDisruptionInjector) applyOperations() error {
+			watcher.podsWithoutIPs = append(watcher.podsWithoutIPs[:podToCreateIdx], watcher.podsWithoutIPs[podToCreateIdx+1:]...)
+		}
+	case watch.Deleted:
+		watcher.tcFiltersFromPodEndpoints, err = i.removeServiceFiltersInList(watcher.tcFiltersFromPodEndpoints, tcFiltersFromPod)
+		if err != nil {
+			return err
+		}
+	}
+
+	if err := i.config.Netns.Exit(); err != nil {
+		return fmt.Errorf("unable to exit the given container network namespace: %w", err)
+	}
+
+	return nil
+}
+
+// watchServiceChanges for every changes happening in the kubernetes service destination or in the pods related to the kubernetes service destination, we update the tc service filters
+func (i *networkDisruptionInjector) watchServiceChanges(watcher serviceWatcher, interfaces []string, flowid string) error {
+	for {
+		// We create the watcher channels when it's closed
+		if watcher.kubernetesServiceWatcher == nil {
+			serviceWatcher, err := i.config.K8sClient.CoreV1().Services(watcher.watchedServiceSpec.Namespace).Watch(context.Background(), metav1.ListOptions{})
+			if err != nil {
+				return fmt.Errorf("error watching the changes for the given kubernetes service (%s/%s): %w", watcher.watchedServiceSpec.Namespace, watcher.watchedServiceSpec.Name, err)
+			}
+
+			i.config.Log.Infow("starting kubernetes service watch", "serviceName", watcher.watchedServiceSpec.Name)
+			watcher.kubernetesServiceWatcher = serviceWatcher.ResultChan()
+		}
+
+		if watcher.kubernetesPodEndpointsWatcher == nil {
+			podsWatcher, err := i.config.K8sClient.CoreV1().Pods(watcher.watchedServiceSpec.Namespace).Watch(context.Background(), metav1.ListOptions{
+				LabelSelector: watcher.labelServiceSelector,
+			})
+			if err != nil {
+				return fmt.Errorf("error watching the list of pods for the given kubernetes service (%s/%s): %w", watcher.watchedServiceSpec.Namespace, watcher.watchedServiceSpec.Name, err)
+			}
+
+			i.config.Log.Infow("starting kubernetes pods watch", "serviceName", watcher.watchedServiceSpec.Name)
+			watcher.kubernetesPodEndpointsWatcher = podsWatcher.ResultChan()
+		}
+
+		select {
+		case state := <-i.config.State.State: // We stop watching when the injection is cleaned
+			if state == Cleaned {
+				return nil
+			}
+		case event, ok := <-watcher.kubernetesServiceWatcher: // We have changes in the service watched
+			if !ok { // channel is closed
+				watcher.kubernetesServiceWatcher = nil
+			} else {
+				if err := i.handleKubernetesServiceChanges(event, &watcher, interfaces, flowid); err != nil {
+					return fmt.Errorf("couldn't apply changes to tc filters: %w", err)
+				}
+			}
+		case event, ok := <-watcher.kubernetesPodEndpointsWatcher: // We have changes in the service watched
+			if !ok { // channel is closed
+				watcher.kubernetesPodEndpointsWatcher = nil
+			} else {
+				if err := i.handleKubernetesPodsChanges(event, &watcher, interfaces, flowid); err != nil {
+					return fmt.Errorf("couldn't apply changes to tc filters: %w", err)
```

Comment:
Hmm, in the case of an error, should we try deleting and recreating the watcher, here?